Readme.md

# AJAX Button

Simple 1 state ajax button.

## Installation

Install with [component(1)](http://component.io):

```sh
$ component install chameleonui/ajax-button
```

If you want to `build example`, you will need to have stylus and chameleon-stylus-plugin installed.

```sh
$ npm install stylus -g     // install stylus globaly
$ npm install               // install package dependencies
``` 

## API

```js
new AjaxButton([options]);
```

```html
<a href="http://experiments.apiary.io/ajax-button" class="button-primary js-ajax-button"><i class="fa fa-check"></i> Send data</a>
```

For now, more you can learn from `example.html`.


### Options

```jss
method: 'POST',
contentType: 'application/x-www-form-urlencoded; charset=UTF-8',
data: function() {
    return window.encodeURIComponent(document.querySelectorAll('#data-for-post')[0].name) + '=' + window.encodeURIComponent(document.querySelectorAll('#data-for-post')[0].value);
},
callback: function (xhrResponse) {
    return xhrResponse.status >= 200 && xhrResponse.status <= 399;
},
states: {
    stateDefault: {
        content: '<i class="fa fa-check"></i> Send data',
        classes: 'button-primary',
        title: 'State Default'
    },
    XHRrequest: {
        content: '<i class="fa fa-spinner fa-spin"></i>',
        classes: 'button-disabled',
        title: 'Loading ...'
    }
}
```

* **data** - Function for geting actual data, called everytime when sending xhr request.
* **callback** - Function returning boolean true|false. xhrResponse is full response, xhrResponse.status is returned status


## ToDo

Write complete readme


## Author(s)

[Edgedesign s.r.o.](http://www.edgedesing.cz) – [Daniel Sitek](https://github.com/danielsitek)

## License

The MIT License (MIT)

Copyright © 2013 Edgedesign s.r.o.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.

History.md
## v1

### 1.0.2

* Fixed jquery component dependency.
* Fixed link to apiary.io in example.html

### 1.0.1

* Found and fixed bug with cached content data. Added data function to options, which will run everytime when you send AJAX request to get and send actualized data.

### 1.0.0

* First public release

index.js
/**
 * 
 * AJAX Button
 * 
 */

var $ = require('jquery');
var Emitter = require('emitter');

var defaults = {
    method: 'POST',
    contentType: 'application/x-www-form-urlencoded; charset=UTF-8',
    data: function() {
        return window.encodeURIComponent('data');
    },
    callback: function (xhrResponse) {
        return xhrResponse.status >= 200 && xhrResponse.status <= 399;
    },
    states: {
        stateDefault: {
            content: '<i class="fa fa-check"></i> Send data',
            classes: 'button-primary',
            title: 'State Default'
        },
        XHRrequest: {
            content: '<i class="fa fa-spinner fa-spin"></i>',
            classes: 'button-disabled',
            title: 'Loading ...'
        }
    }
};

function AjaxButton(options) {
    this.options = options || {};
    for (var i in defaults) {
        if (!(this.options[i])) this.options[i] = defaults[i];
    }
    this.response = {};
    this.self;
    this.active = 0;
    // this.postData = {};
}

module.exports = AjaxButton;

Emitter(AjaxButton.prototype);

AjaxButton.prototype.init = function(self) {
    
    var component = this;
    component.self = self;

    component._stateChange('XHRrequest');
};

AjaxButton.prototype._click = function(fn) {
    var component = this;

    $(component.self).on('click.ajaxpost', function(event) {

        event.preventDefault();
        fn(event);
    });
};

AjaxButton.prototype._clearEvent = function() {
    
    var component = this;

    $(component.self).off('click.ajaxpost');
};

AjaxButton.prototype._stateChange = function(newStateName) {
    
    var component = this;

    // clear binded event - prevents multiple clicks
    component._clearEvent();

    // replace full class list
    if (component.options.states[newStateName]) {

        component.self.setAttribute('class', component.options.states[newStateName].classes);
        component.self.setAttribute('title', component.options.states[newStateName].title);
        component.self.innerHTML = component.options.states[newStateName].content;

        component[newStateName]();

    } else {

        console.error('State: ' + newStateName + ' does not exist');
    }
};

AjaxButton.prototype.stateDefault = function() {

    var component = this;

    component._click(function(){

        component._stateChange('XHRrequest');

    });
};

AjaxButton.prototype.XHRrequest = function() {
    
    var component = this;

    component._click(function(){});

    component.active = 1;

    component.emit('request');

    component._xhr(component.self.getAttribute('href'), component.options.data(), component.options.contentType, component.options.method, function(xhr){

        component.response = xhr;
        component._XHRresponse();

    });
};

AjaxButton.prototype._XHRresponse = function() {
    
    var component = this;

    component.active = 0;
    
    component._stateChange('stateDefault');

    if ( component.options.callback(component.response) ) {

        component.emit('success');

    } else {

        component.emit('error');
    }

    component.emit('response');
};

AjaxButton.prototype._xhr = function(setUrl, data, contentType, method, resFn) {

    $.ajax({
        type: method,
        url: setUrl,
        data: data,
        contentType: contentType
    })
    .done(function(data, textStatus, jqXHR) {
        resFn(jqXHR);
    })
    .fail(function(jqXHR, textStatus, errorThrown) {
        resFn(jqXHR);
    });
};
